suite/src/benchmark.rs

```diff
@@ -0,0 +1,75 @@
+use core::arch::asm;
+
+/// Returns the machine cycle counter
+#[inline]
+pub fn get_cycle() -> u64 {
+    let counter_lo: u32;
+    let counter_hi: u32;
+
+    unsafe {
+        asm!(
+            "1: csrr {hi_old}, mcycleh",
+            "csrr {lo}, mcycle",
+            "csrr {hi}, mcycleh",
+            "bne {hi_old}, {hi}, 1b",
+            hi_old = out(reg) _,
+            lo = out(reg) counter_lo,
+            hi = out(reg) counter_hi
+        );
+    }
+    ((counter_hi as u64) << 32u64) + counter_lo as u64
+}
+
+pub mod examples {
+    use crate::platform::{self, Platform};
+
+    pub fn sha256_benchmark() {
+        if let Some(hmac_module) = platform::current().get_sha256_module() {
+            let input = [
+                0xdf3f6198, 0x04a92fdb, 0x4057192d, 0xc43dd748, 0xea778adc, 0x52bc498c, 0xe80524c0,
+                0x14b81119, 0xdf3f6198, 0x04a92fdb, 0x4057192d, 0xc43dd748, 0xea778adc, 0x52bc498c,
+                0xe80524c0, 0x14b81119, 0xdf3f6198, 0x04a92fdb, 0x4057192d, 0xc43dd748, 0xea778adc,
+                0x52bc498c, 0xe80524c0, 0x14b81119, 0xdf3f6198, 0x04a92fdb, 0x4057192d, 0xc43dd748,
+                0xea778adc, 0x52bc498c, 0xe80524c0, 0x14b81119, 0xdf3f6198, 0x04a92fdb, 0x4057192d,
+                0xc43dd748, 0xea778adc, 0x52bc498c, 0xe80524c0, 0x14b81119, 0xdf3f6198, 0x04a92fdb,
+                0x4057192d, 0xc43dd748, 0xea778adc, 0x52bc498c, 0xe80524c0, 0x14b81119, 0xdf3f6198,
+                0x04a92fdb, 0x4057192d, 0xc43dd748, 0xea778adc, 0x52bc498c, 0xe80524c0, 0x14b81119,
+                0xdf3f6198, 0x04a92fdb, 0x4057192d, 0xc43dd748, 0xea778adc, 0x52bc498c, 0xe80524c0,
+                0x14b81119,
+            ];
+            let mut output = [0u32; 8];
+
+            let cycle1 = super::get_cycle();
+            hmac_module.init_sha256();
+            let cycle2 = super::get_cycle();
+            hmac_module.write_input(&input);
+            let cycle3 = super::get_cycle();
+            hmac_module.wait_for_completion();
+            let cycle4 = super::get_cycle();
+            hmac_module.read_digest(&mut output);
+            let cycle5 = super::get_cycle();
+
+            assert_eq!(
+                output,
+                [
+                    // precomputed by sha2 crate
+                    0xa24ef743, 0xed238e92, 0x8f5fe495, 0x7959a1fa, 0x06b1d250, 0x147ed98d,
+                    0xd817e3b2, 0xb32854ae,
+                ]
+            );
+
+            let init_runtime = cycle2 - cycle1;
+            let write_input_runtime = cycle3 - cycle2;
+            let wait_for_completion_runtime = cycle4 - cycle3;
+            let read_output_runtime = cycle5 - cycle4;
+
+            println!();
```

Comment:
Try to make this configurable according to the input parameter, i.e the number of benchmark runs. For simplicity you can remove the names, only output values. The names can be added by the function that calls this one:

```
initialization,writing_input,computation,reading_output
```

Just output one row of values for each run, followed by a newline. Include a comment with the names of the values. What do you think?